pyomo/contrib/incidence_analysis/interface.py

```diff
@@ -507,17 +710,55 @@ def get_diagonal_blocks(self, variables=None, constraints=None):
         return block_vars, block_cons
 
     def dulmage_mendelsohn(self, variables=None, constraints=None):
-        """
-        Returns the Dulmage-Mendelsohn partition of the incidence graph
-        of the provided variables and constraints.
+        """Partition variables and constraints according to the Dulmage-
+        Mendelsohn characterization of the incidence graph
 
-        Returns:
-        --------
-        ColPartition namedtuple and RowPartition namedtuple.
-        The ColPartition is returned first to match the order of variables
-        and constraints in the method arguments.
-        These partition variables (columns) and constraints (rows)
-        into overconstrained, underconstrained, unmatched, and square.
+        Variables are partitioned into the following subsets:
+
+        - **unmatched** - Variables not matched in a particular maximum
+          cardinality matching
+        - **underconstrained** - Variables that *could possibly be* unmatched
+          in a maximum cardinality matching
+        - **square** - Variables in the well-constrained subsystem
+        - **overconstrained** - Variables matched with constraints that can
+          possibly be unmatched
+
+        Constraints are partitioned into the following subsets:
+
+        - **underconstrained** - Constraints matched with variables that can
+          possibly be unmatched
+        - **square** - Constraints in the well-constrained subsystem
+        - **overconstrained** - Constraints that *can possibly be* unmatched
+          with a maximum cardinality matching
+        - **unmatched** - Constraints that were not matched in a particular
+          maximum cardinality matching
+
+        Returns
+        -------
+        var_partition: ``ColPartition`` named tuple
+            Partitions variables into square, underconstrained, overconstrained,
+            and unmatched.
+        con_partition: ``RowPartition`` named tuple
+            Partitions constraints into square, underconstrained,
+            overconstrained, and unmatched.
+
+        Example
+        -------
+        >>> import pyomo.environ as pyo 
+        >>> from pyomo.contrib.incidence_analysis import IncidenceGraphInterface
+        >>> m = pyo.ConcreteModel()
+        >>> m.x = pyo.Var([1, 2]) 
+        >>> m.eq1 = pyo.Constraint(expr=m.x[1]**2 == 7)
+        >>> m.eq2 = pyo.Constraint(expr=m.x[1]*m.x[2] == 3)
+        >>> m.eq3 = pyo.Constraint(expr=m.x[1] + 2*m.x[2] == 5)
+        >>> igraph = IncidenceGraphInterface(m)
+        >>> var_dmp, con_dmp = igraph.dulmage_mendelsohn()
+        >>> print([v.name for v in var_dmp.overconstrained])
+        ['x[1]', 'x[2]']
+        >>> print([c.name for c in con_dmp.overconstrained])
+        ['eq1', 'eq2']
+        >>> print([c.name for c in con_dmp.unmatched])
```

Comment:
You should add the `.. doctest::` command in this docstring so that this short example is tested (as long as this docstring is included in API documentation on RTD)